consensus/src/pos/target.rs

```diff
@@ -0,0 +1,567 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chainstate_types::{BlockIndexHandle, BlockIndexHistoryIterator};
+use common::{
+    chain::{
+        block::ConsensusData, Block, ChainConfig, GenBlock, GenBlockId, PoSChainConfig, PoSStatus,
+        RequiredConsensus,
+    },
+    primitives::{BlockHeight, Compact, Id},
+    Uint256,
+};
+use itertools::Itertools;
+use utils::ensure;
+
+use crate::pos::error::ConsensusPoSError;
+
+fn calculate_average_block_time(
+    chain_config: &ChainConfig,
+    pos_config: &PoSChainConfig,
+    block_id: Id<Block>,
+    block_height: BlockHeight,
+    block_index_handle: &impl BlockIndexHandle,
+) -> Result<u64, ConsensusPoSError> {
+    let history_iter = BlockIndexHistoryIterator::new(block_id.into(), block_index_handle);
+
+    // Determine net upgrade version that current block belongs to.
+    // Across versions config parameters might change so it's invalid to mix versions for target calculations
+    let (_, net_version) = chain_config
+        .net_upgrade()
+        .version_at_height(block_height)
+        .expect("NetUpgrade must've been initialized");
+    let net_version_range = chain_config
+        .net_upgrade()
+        .height_range(net_version)
+        .expect("NetUpgrade must've been initialized");
+
+    // Get timestamps from the history but make sure they belong to the same consensus version.
+    // Then calculate differences of adjacent elements.
+    let block_diffs = history_iter
+        .take(pos_config.block_count_to_average_for_blocktime())
+        .filter(|block_index| net_version_range.contains(&block_index.block_height()))
+        .map(|block_index| block_index.block_timestamp().as_int_seconds())
+        .tuple_windows::<(u64, u64)>()
+        .map(|t| t.0 - t.1);
```

Comment:
I think the `tuple_windows` and `map` will, when finally summed, give you just the difference between the first and the last element. Could it be implemented easily without taking differences of adjacent elements? Maybe it does not matter much, and I don't immediately see a nice solution that works on iterators.

`(x0 - x1) + (x1 - x2) + (x2 - x3) + ... (xk - xn)` = `x0 + (-x1 + x1) + (-x2 + x2) + ... - xn` = `x0 + 0 + 0 + ... - xn` = `x0 - xn`